File: Python/Reverse Words in a String.py
"""
Given an input string, reverse the string word by word.
For example,
Given s = "the sky is blue",
return "blue is sky the".

Clarification:
What constitutes a word?
A sequence of non-space characters constitutes a word.
Could the input string contain leading or trailing spaces?
Yes. However, your reversed string should not contain leading or trailing spaces.
How about multiple spaces between two words?
Reduce them to a single space in the reversed string.
"""

class Solution:
    # @param s, a string
    # @return a string
    def reverseWords(self, s):
        res = ''
        word = ''
        for char in s:
            if char != ' ':
                word += char
            elif len(word) > 0:
                if res != '':
                    res = ' ' + res
                res = word + res
                word = ''

        if len(word) > 0:
            if res != '':
                res = ' ' + res
            res = word + res
        return res

File: Java/Reverse Words in a String.java
/*
Given an input string, reverse the string word by word.

For example,
Given s = "the sky is blue",
return "blue is sky the".
*/

public class Solution {
    public String reverseWords(String s) {
        int start = 0;
        int end = s.length() - 1;
        while(start< end && s.charAt(start) == ' ') start++;
        while(start< end && s.charAt(end) == ' ' ) end--;
        StringBuilder result = new StringBuilder();
        int p1 = end + 1;
        for(int k = end;k>=start;k--)
        {
            if(s.charAt(k) == ' ')
            {
                p1 = k;
            }
            else
            {
                if(k == start || s.charAt(k-1) == ' ')
                {
                    if(result.length() != 0){
                        result.append(' ');
                    }
                    result.append(s.substring(k,p1));
                }
            }
        }
        return result.toString();
    }
}